### PR TITLE
Farewell TrainableDropout

### PR DESCRIPTION
**Pull request by codemzs**
Dropout-12 (and onwards) has a training mode that makes TrainableDropout obsolete.

### CODE REVIEW

File: orttraining/orttraining/test/optimizer/graph_transform_test.cc [hunk 1]
```diff
@@ -97,7 +96,6 @@ TEST_F(GraphTransformationTests, BiasDropoutFusionTest) {
   TestBiasDropoutFusion(MODEL_FOLDER "fusion/bias_dropout_residual_fusion1.onnx", *logger_);
   TestBiasDropoutFusion(MODEL_FOLDER "fusion/bias_dropout_residual_fusion2.onnx", *logger_);
   TestBiasDropoutFusion(MODEL_FOLDER "fusion/bias_dropout_residual_fusion_mismatch.onnx", *logger_, 1);
-  TestBiasDropoutFusion(MODEL_FOLDER "fusion/bias_trainabledropout_residual_fusion.onnx", *logger_);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/optimizer/graph_transform_test.cc:100]: the bias_trainabledropout_residual_fusion.onnx can also be deleted.

Comment (SherlockNoMad) [reply orttraining/orttraining/test/optimizer/graph_transform_test.cc:100]: also update the python script that generate this .onnx file.

**Comment by kit1980** [comment]
This means that ONNX opsets before 12 are not supported anymore for training, right?
So basically it's not just deprecate TrainableDropout, but deprecate opset 10 for training.
Please include in this PR some TODOs for code and docs to clean up the rest of opset 10 (both backend and frontend, in places like this https://github.com/microsoft/onnxruntime/blob/4d517c68a3958b97dcfe8267714ef0af130eced6/orttraining/orttraining/python/training/orttrainer_options.py#L197)

**Comment by codemzs** [comment]
Hi Sergii, the goal of this PR is just to deprecate trainable dropout since Dropout-12 has a training mode, as for opset-10 not being used for training I'm not aware of this but definitely this is not within the scope of this PR and may be Sherlock can clarify that.

---
In reply to: [726417828](https://github.com/microsoft/onnxruntime/pull/5793#issuecomment-726417828) [](ancestors = 726417828)

**Comment by SherlockNoMad** [comment]
> This means that ONNX opsets before 12 are not supported anymore for training, right?
> So basically it's not just deprecate TrainableDropout, but deprecate opset 10 for training.
> Please include in this PR some TODOs for code and docs to clean up the rest of opset 10 (both backend and frontend, in places like this
> 
> https://github.com/microsoft/onnxruntime/blob/4d517c68a3958b97dcfe8267714ef0af130eced6/orttraining/orttraining/python/training/orttrainer_options.py#L197
> 
> )

Sync with Sergii offline, deprecating opset 10 will come in anther PR.

**Review by SherlockNoMad — APPROVED** [review]

**Comment by SherlockNoMad** [comment]
I pushed a commit to addressed to my own comments. Let's merge this.

File: onnxruntime/test/python/onnxruntime_test_ort_trainer.py [hunk 2]
```diff
@@ -652,8 +652,8 @@ def testMNISTTrainingCheckpoint(self):
             assert np.array_equal(state_dict[key], loaded_state_dict[key])
 
     def testBertTrainingBasic(self):
-        expected_losses = [11.034271, 11.125311, 11.006095, 11.046938, 11.027476, 11.015745, 11.060884, 10.971851]
-        expected_eval_loss = [10.95898914]
+        expected_losses = [11.027887, 11.108191, 11.055356, 11.040912, 10.960277, 11.02691, 11.082471, 10.920979]
```
[the remark this answers is not in the document]
Comment (codemzs) [reply onnxruntime/test/python/onnxruntime_test_ort_trainer.py:655]: do we need to address this? I haven't investigated why the values are different due to memory investigation work and this is the reason I had not pushed another commit.

---
In reply to: [522516168](https://github.com/microsoft/onnxruntime/pull/5793#discussion_r522516168) [](ancestors = 522516168)

**Review by thiagocrepaldi — APPROVED** [review]

**Comment by kit1980** [comment]
Now testBertTrainingGradientAccumulation fails with 
onnxruntime.capi.onnxruntime_pybind11_state.InvalidGraph: [ONNXRuntimeError] : 10 : INVALID_GRAPH : This is an invalid model. Error in Node:TrainableDropout_114 : No Op registered for TrainableDropout with domain_version of 10

**Comment by kit1980** [comment]
> Now testBertTrainingGradientAccumulation fails with
> onnxruntime.capi.onnxruntime_pybind11_state.InvalidGraph: [ONNXRuntimeError] : 10 : INVALID_GRAPH : This is an invalid model. Error in Node:TrainableDropout_114 : No Op registered for TrainableDropout with domain_version of 10

onnxruntime/test/testdata/bert_toy_postprocessed.onnx needs to be updated to opset 12.

**Comment by kit1980** [comment]
I uploaded opset 12 bert_toy_postprocessed.onnx from https://github.com/microsoft/onnxruntime/blob/6eb5549cb976e34efd75e32c2e20e21beddf01c8/onnxruntime/test/testdata/bert_toy_postprocessed.onnx

File: orttraining/orttraining/training_ops/cuda/nn/dropout.cc [hunk 3]
```diff
@@ -9,6 +9,17 @@
...
+#define ALL_IEEE_FLOAT_TENSOR_TYPES {DataTypeImpl::GetTensorType<float>(),      \
+                                     DataTypeImpl::GetTensorType<double>(),     \
+                                     DataTypeImpl::GetTensorType<MLFloat16>(),  \
+                                     DataTypeImpl::GetTensorType<BFloat16>()}
+#define ALL_IEEE_FLOAT_DATA_TYPES float, MLFloat16, double, BFloat16
+#else
+#define ALL_IEEE_FLOAT_TENSOR_TYPES DataTypeImpl::AllIEEEFloatTensorTypes()
+#define ALL_IEEE_FLOAT_DATA_TYPES float, MLFloat16, double
+#endif
+
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/training_ops/cuda/nn/dropout.cc]: Should we refactor https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/framework/data_types.cc#L898-L905 to prevent code duplicate or implementation discrepancies?

Comment (kit1980) [reply orttraining/orttraining/training_ops/cuda/nn/dropout.cc]: This is just merged from master, not related to this PR.

**Review by SherlockNoMad — APPROVED** [review]

**Review by thiagocrepaldi — APPROVED** [review]
